### ISSUE TITLE
NewRelic Sink

### ISSUE THREAD

**Issue by johnkattenhorn** [serilog/serilog#546]
We've been asked to produce an integration between our platform and a clients DevOps platform NewRelic. 

I've used the NewRelic API before and don't think it would be too hard to get something working. Ideally the client would like us to support to transferring the metrics we get from Serilog.Metrics into NewRelic as metrics not messages but not sure how best to do this so we will leave till last.

Timeframe for this work is this week for demo next, we are want to release it back to the community so any feedback on features or ideas around approach would be most welcome.

So far I've copied the plumping and overall architecture from the Serilog.Sinks.Seq sink from @nblumhardt as we liked the implementation an now just need to map to the NewRelic API client.

**Comment by merbla** [comment]
TBH I think the Seq sink has some great smarts. Keen to see how you go as I have recently updated the Splunk HTTP Sink and I have a few changes yet to commit to make it more durable.  The other sink that may be worth a look would be (Elastic Search)[https://github.com/serilog/serilog-sinks-elasticsearch]

Is NewRelic primarily a HTTP endpoint?

**Comment by johnkattenhorn** [comment]
I'll take a look at both at Splunk and Elastic search as well see if there's something to be learned, thanks for the suggestion.

It looks from the peeking inside the source it's using Trace.Write which suggest there must by aa TraceWriter involved somewhere that they have built ? I'm sure you could side-step their entire API .dll and go straight to an Http API.

For reference here is the NewRelic API documentation 

https://docs.newrelic.com/docs/agents/net-agent/features/net-agent-api

I'll but the code into my own repo for comment and then find out how I should conform to the Sink Repo standards etc if it's good enough to be accepted ;-)

**Comment by runxc1** [comment]
So just came across this and was thinking it would be nice to have a sync for NewRelic.  Did you make it anywhere?

**Comment by johnkattenhorn** [comment]
I've just uploaded the code we created:

https://github.com/Applicita/serilog-sinks-newrelic

We started with a fork of the Serilog.Sinks.Seq but stopped short of implementing the durable side of things as it was for a client demo originally.

We also attempted to implement an idea to push metrics from [Serilog.Metrics](https://github.com/serilog-metrics/serilog-metrics) into NewRelic but we were still learning NewRelic and discovered you needed a custom transaction to send custom metric and you could only do this by implementing custom method discovering which looked to us like might be a sort of duplicate.

On a wider note as a small software house we are running more and more into a natural friction when our clients wants us to use say Stackify or NewRelic to provide Logging + APM. Serilog log sinks have served us well in being able to abstract the logging away from the Logging Collector but not so much on the APM side of things, @nblumhardt is this something you've seen or thought about much ?

**Comment by nblumhardt** [comment]
@johnkattenhorn I've experimented a bit in the metrics space via https://github.com/nblumhardt/influxdb-lineprotocol. The two things - logs and metrics - seem so similar on the outside, but I've come to the conclusion they're fundamentally different beasts. Interested to see how things evolve in this space :-)

**Comment by johnkattenhorn** [comment]
@nblumhardt - Let me know if we should take this conversation elsewhere :-)

I thought the same, they look similar on the surface but they are really different. In the both APM+ I've looked at which is Stackify and NewRelic they both take a similar approach which is to intecept calls to methods and measuring their performance by looking at the CLR and using either XML (NewRelic) or Json (Stackify) file to control that process.

Having used SerilogMetrics originally I thought that wasn't very nice as I preferred the code approach but now having over 200+ Metric usings in my code base and having to redeploy when I want to adjust the instrumentation I'm moving towards the CLR / File approach.

I saw this project for [NewRelic](http://pablissimo.github.io/nrconfig/) which uses attributes to generate the NewRelic file and I did think about using this idea to provide a degree of abstraction that would allow to create sinks for different APM's which would generate the specific file.

I need to look at some more APM system to see if they are all doing there performance tracking this way before I decided if it's worth building.
